mirage/tar_mirage.ml

```diff
@@ -360,33 +397,37 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
           div data_start_bytes sector_size,
           rem data_start_bytes sector_size
         in
-        let end_sector = div end_bytes sector_size in
-        let end_sector_offset = rem end_bytes sector_size in
+        let end_sector = div (add end_bytes sector_size) sector_size in
+        let last_sector_offset = rem end_bytes sector_size in
         let pad = Tar.Header.compute_zero_padding_length hdr in
 
-        (* Read slack for last sector *)
-        begin
-          if end_sector_offset = 0L then
-            Lwt.return_ok Cstruct.empty
-          else begin
-            let buf = Cstruct.create (to_int sector_size) in
-            BLOCK.read t.b end_sector [buf] >|= function
-            | Error e -> Error (`Block e)
-            | Ok () -> Ok (snd (Cstruct.split buf (to_int end_sector_offset)))
-          end
-        end >>>= fun slack ->
-        let data = Cstruct.concat [ data; Cstruct.create (pad + to_int sentinel); slack ] in
-        let first_sector, remaining_sectors =
-          let s =
-            Stdlib.min
-              (Cstruct.length data)
-              (to_int (sub sector_size data_start_sector_offset))
-          in
-          Cstruct.split data s
+        let data = Cstruct.concat [
+            Cstruct.create (to_int data_start_sector_offset);
+            data;
+            Cstruct.create (pad + to_int sentinel);
+            (if last_sector_offset = 0L then
+               Cstruct.empty
+             else
+               Cstruct.create (to_int (sub sector_size last_sector_offset)));
+          ]
         in
+        (* [data] is always at least one sector as the sentinel is always present *)
+        let first_sector, remaining_sectors = Cstruct.split data t.info.sector_size in
+        let last_sector =
+          (* sub on whole [data] as the first sector and last sector might be the same *)
+          Cstruct.sub data
+            (Stdlib.max 0 (Cstruct.length data - t.info.sector_size))
+            t.info.sector_size
+        in
+        (* blit in slack at the end *)
+        read_partial_sector t (max data_start_sector (pred end_sector))
+          ~offset:last_sector_offset last_sector
+          ~length:(sub sector_size last_sector_offset) >>>= fun () ->
         (* to write robustly as we can:
-           - we write tar blocks 2..end,
-           - then the header AND the first tar block
+           - we write sectors 2..n,
+           - then the header,
+           - then we blit the first (data) sector as it may contain the header,
+           - finally we write the first (data) sector which contains the first tar data block.
```

Comment:
Perhaps we should explain the relation between sectors and blocks? I'm not sure how to phrase it.